### DISCUSSION TITLE
What are the steps to build a new docker container?

### DISCUSSION THREAD

**Discussion by jamesnos** [flipt-io/flipt#491] [Q&A] [answered]
Hi All,
Just needing a bit of a point in the right direction.

I'm just trying to bake my own docker image locally of the ui and server together (binary). Unable to do so and tried multiple different ways. 

Do you have some steps? How would your CICD process build the docker image E2E from a cold start.

**Comment by markphelps** [comment, marked as answer]
Hi @jamesnos 

I **just** made some changes to how Flipt is built/bootstrapped, but to answer your question you should be able to build your own image two different ways.

## Build an image as close to the official Flipt image as possible

**mac os/homebrew required**

This is how I build and release the Flipt docker image, using a combo of [musl-cross](https://github.com/richfelker/musl-cross-make) and [goreleaser](https://goreleaser.com/intro/)

Basically the process is:

1. Flipt binary is built using go-releaser, which has some [special configs](https://github.com/markphelps/flipt/blob/1a797d2a4b27da11b8072a074e4f3f5d1ee3aa01/.goreleaser.yml#L1-L15) to build for Linux, using cgo (required for the  https://github.com/mattn/go-sqlite3 lib)
2. go-releaser copies the resulting binary (now built for linux/amd64) into the [build/Dockerfile](https://github.com/markphelps/flipt/blob/master/build/Dockerfile), this is the Dockerfile that actually results in the release image, copying configs, migrations, etc

To do this you'll need (assuming you are on a mac):

1. [goreleaser](https://goreleaser.com/install/)
2. [homebrew-musl-cross](https://github.com/FiloSottile/homebrew-musl-cross)

Once those are installed, you should just need to run `make snapshot`, which will then build a snapshot build (which is the same as a release, it just doesnt create a release on Github/etc)

This should output the name of the snapshot image at the end, then you can just run it just like [the docs say](https://flipt.io/docs/installation/#run-the-image) but with your image name instead.

## Build an image based on the dev Docker file

This way will not give you the exact same result as the step above (ie it will be a larger image than the release version), but will work if you just want to explore/run Flipt but want to build and run it within Docker.

1. Modify the [Dockerfile](https://github.com/markphelps/flipt/blob/master/Dockerfile) and add the following two lines at the **bottom**:
```
RUN make build
CMD [ "./bin/flipt", "--config", "./config/local.yml" ]
```
2. Build an image using this new Dockerfile:
```
docker build -t flipt:dev . 
```
3. Run according to [the docs](https://flipt.io/docs/installation/#run-the-image) but with your `flipt:dev` instead of `markphelps/flipt`

I hope this helps!

**Reply by jamesnos** [reply]
Thanks @markphelps , this is a huge help, appreciate the detailed response. Really impressed with the project overall. Well done!
